File: src/my_project/easy_problems/from201to250/flood_fill.py
from typing import List 

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        
        m = len(image)
        n = len(image[0])
        oldColor = image[sr][sc]
        visited = []
        #visited array keeps a track of visited elements. without this maximum recursion depth 
        #exceeded error might be encountered
        
        def floodfilluntil(image,x,y,newColor,oldColor):
            if (x,y) not in visited:
                visited.append((x,y))
            else:
                return
            if x<0 or y<0 or x>=m or y>=n:
                return
            if image[x][y]!=oldColor:
                return
            
            image[x][y] = newColor
            
            floodfilluntil(image,x+1,y,newColor,oldColor)
            floodfilluntil(image,x-1,y,newColor,oldColor)
            floodfilluntil(image,x,y+1,newColor,oldColor)
            floodfilluntil(image,x,y-1,newColor,oldColor)
            
            return image
        
        return(floodfilluntil(image,sr,sc,newColor,oldColor))
        

File: my_project_typescript/easy_problems_typescript/from101to150/floodFill.ts
function floodFill(image: number[][], sr: number, sc: number, newColor: number): number[][] {

    let m: number = image.length
    let n: number = image[0].length
    let oldColor: number = image[sr][sc]
    // visitedCoordinates containes the coordinates of the elements visited 
    let visitedCoordinates: Array<Array<number>> =[]

    function floodFillUntil(image: number[][], x: number, y: number, newColor: number, oldColor: number): number[][]{
        if (!visitedCoordinates.includes([x,y])){
            visitedCoordinates.push([x,y])
        }
        else{
            return image
        }
        if (x <0 || y < 0 || x >= m || y >= n){
            return image
        }
        if (image[x][y] != oldColor){
            return image
        }

        image[x][y] = newColor
        floodFillUntil(image,x-1, y, newColor, oldColor)
        floodFillUntil(image,x+1, y, newColor, oldColor)
        floodFillUntil(image,x, y-1, newColor, oldColor)
        floodFillUntil(image,x, y+1, newColor, oldColor)


        return image
    }

    return floodFillUntil(image, sr,sc,newColor, oldColor)
};

console.log(floodFill([[1,1,1],[1,1,0],[1,0,1]],
    1,
    1,
    2))